### PR TITLE
[MXNET-531] improvement of adding source.jar to assembly

### PR DESCRIPTION
**Pull request by lanking520**
## Description ##
This is an update on the `assembly` side adding source code jar. It will sure help users to get documentation of apis when they are using the Scala package. Currently, it also add `Javadoc` jar to assembly but only supporting single module
#11618
@nswamy @yzhliu @andrewfayres 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: scala-package/assembly/src/source.xml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  <formats>
+    <format>jar</format>
+  </formats>
+  <includeBaseDirectory>false</includeBaseDirectory>
+  <moduleSets>
+    <moduleSet>
+      <useAllReactorProjects>true</useAllReactorProjects>
+      <includes>
+        <!--Please put your module defs here-->
+        <include>org.apache.mxnet:mxnet-core_${scala.binary.version}</include>
```
Comment (nswamy) [inline scala-package/assembly/src/source.xml]: note to self: add infer package after pom-flle updates for maven publish

**Comment by lanking520** [comment]
@gigasquid I tried to reproduce the issue in CI but failed on my Ubuntu Machine
```
lein test
+ lein test
Could not find artifact org.apache.mxnet:mxnet-full_2.11-linux-x86_64-cpu:jar:1.3.0-SNAPSHOT in clojars (https://repo.clojars.org/)
Could not find artifact org.apache.mxnet:mxnet-full_2.11-linux-x86_64-cpu:jar:1.3.0-SNAPSHOT in staging (https://repository.apache.org/content/repositories/staging)
This could be due to a typo in :dependencies, file system permissions, or network issues.
If you are behind a proxy, try setting the 'http_proxy' environment variable.
```
Could you please help with the `Clojure` package build failure, I have `.so` generated here:
```
/home/ubuntu/incubator-mxnet/scala-package/native/linux-x86_64-cpu/target/libmxnet-scala-linux-x86_64-cpu.so
```
`-Djava.library.path=[lib path]`

**Comment by benkamphaus** [comment]
@nswamy and @lanking520 -- (and cc @gigasquid ) I think something may be going on with the scala build based on some of the assembly changes that impacts the maven artifact of which Clojure is a consumer. Since I'm new to the build process here and reading through CI, it's possible I could be misinterpreting something though.

In the CI logs I see this set of warnings:

```
[INFO] Reading assembly descriptor: src/main/assembly/assembly.xml

[WARNING] Artifact: org.apache.mxnet:mxnet-full_2.11-linux-x86_64-cpu:jar:1.3.0-SNAPSHOT references the same file as the assembly destination file. Moving it to a temporary location for inclusion.

[INFO] Building jar: /work/mxnet/scala-package/assembly/linux-x86_64-cpu/target/mxnet-full_2.11-linux-x86_64-cpu-1.3.0-SNAPSHOT.jar

[WARNING] Configuration options: 'appendAssemblyId' is set to false, and 'classifier' is missing.

Instead of attaching the assembly file: /work/mxnet/scala-package/assembly/linux-x86_64-cpu/target/mxnet-full_2.11-linux-x86_64-cpu-1.3.0-SNAPSHOT.jar, it will become the file for main project artifact.

NOTE: If multiple descriptors or descriptor-formats are provided for this project, the value of this file will be non-deterministic!

[WARNING] Replacing pre-existing project main-artifact file: /work/mxnet/scala-package/assembly/linux-x86_64-cpu/target/archive-tmp/mxnet-full_2.11-linux-x86_64-cpu-1.3.0-SNAPSHOT.jar

with assembly file: /work/mxnet/scala-package/assembly/linux-x86_64-cpu/target/mxnet-full_2.11-linux-x86_64-cpu-1.3.0-SNAPSHOT.jar

[INFO] 

[INFO] --- maven-assembly-plugin:2.5.5:single (sources-jar) @ mxnet-full_2.11-linux-x86_64-cpu ---

[INFO] Reading assembly descriptor: /work/mxnet/scala-package/assembly/src/source.xml

[INFO] Processing sources for module project: org.apache.mxnet:mxnet-core_2.11:jar:1.3.0-SNAPSHOT

[INFO] Building jar: /work/mxnet/scala-package/assembly/linux-x86_64-cpu/target/mxnet-full_2.11-linux-x86_64-cpu-1.3.0-SNAPSHOT-sources.jar

[WARNING] Configuration options: 'appendAssemblyId' is set to false, and 'classifier' is missing.

Instead of attaching the assembly file: /work/mxnet/scala-package/assembly/linux-x86_64-cpu/target/mxnet-full_2.11-linux-x86_64-cpu-1.3.0-SNAPSHOT-sources.jar, it will become the file for main project artifact.

NOTE: If multiple descriptors or descriptor-formats are provided for this project, the value of this file will be non-deterministic!

[WARNING] Replacing pre-existing project main-artifact file: /work/mxnet/scala-package/assembly/linux-x86_64-cpu/target/mxnet-full_2.11-linux-x86_64-cpu-1.3.0-SNAPSHOT.jar

with assembly file: /work/mxnet/scala-package/assembly/linux-x86_64-cpu/target/mxnet-full_2.11-linux-x86_64-cpu-1.3.0-SNAPSHOT-sources.jar

[INFO] 

[INFO] --- maven-assembly-plugin:2.5.5:single (javadoc-jar) @ mxnet-full_2.11-linux-x86_64-cpu ---

[INFO] Reading assembly descriptor: /work/mxnet/scala-package/assembly/src/javadoc.xml

[INFO] Building jar: /work/mxnet/scala-package/assembly/linux-x86_64-cpu/target/mxnet-full_2.11-linux-x86_64-cpu-1.3.0-SNAPSHOT-javadoc.jar

[WARNING] Configuration options: 'appendAssemblyId' is set to false, and 'classifier' is missing.

Instead of attaching the assembly file: /work/mxnet/scala-package/assembly/linux-x86_64-cpu/target/mxnet-full_2.11-linux-x86_64-cpu-1.3.0-SNAPSHOT-javadoc.jar, it will become the file for main project artifact.

NOTE: If multiple descriptors or descriptor-formats are provided for this project, the value of this file will be non-deterministic!

[WARNING] Replacing pre-existing project main-artifact file: /work/mxnet/scala-package/assembly/linux-x86_64-cpu/target/mxnet-full_2.11-linux-x86_64-cpu-1.3.0-SNAPSHOT-sources.jar

with assembly file: /work/mxnet/scala-package/assembly/linux-x86_64-cpu/target/mxnet-full_2.11-linux-x86_64-cpu-1.3.0-SNAPSHOT-javadoc.jar

```

Carin looked into this a bit as well and passed me some notes. From this I can see in her logs that she saw the javadoc jar overwrite the project jar in her local maven repo:

```
[INFO] Installing /Users/cmeier/workspace/deep-learning/mxnet/scala-package/assembly/osx-x86_64-cpu/target/mxnet-full_2.11-osx-x86_64-cpu-1.3.0-SNAPSHOT-javadoc.jar to /Users/cmeier/.m2/repository/org/apache/mxnet/mxnet-full_2.11-osx-x86_64-cpu/1.3.0-SNAPSHOT/mxnet-full_2.11-osx-x86_64-cpu-1.3.0-SNAPSHOT.jar
```

I see a [related StackOverflow issue](https://stackoverflow.com/a/38484021) that would suggest that the wrong jar getting published as the main artifact could be the culprit here.

**Comment by lanking520** [comment]
Hi @benkamphaus @gigasquid many thanks for your investigation and it really save my days. 

I think I have screwed up the `make scalainstall` process to let javadoc install and that causing Clojure crashes 100%. I will start looking into the link you have provided and see if there is any changes I should made. Definitely avoid the overwritten artifact!

**Comment by lanking520** [comment]
It finally passed the Clojure test as expected! Thanks guys!

**Pull request merged**